umpf: add new flag to insert 'Upstream-Status: Pending' lines to all patches

michaelolbrich opened the pull request:
Starting with the mickledore release, the upstream status is required for all patches in Yocto.
Add a new flag "upstreamstatus" to handle this. If set to "insert", "Upstream-Status: Pending" is added to the commit message of all patches during "umpf tag".

ejoerns commented:
Looks quite nice already!

For those wondering how to use it, call:

    $ umpf tag --flags upstreamstatus=insert

Maybe the flag usage still needs a bit more description?

One thing that is still missing an "Upstream-Status" is the release version patch itself. This could safely be set to

    Upstream-Status: Inappropriate [autogenerated]

but `Upstream-Status: pending` would fit as well first of all (if easier to implement).

OnkelUlla reviewed (approved):
Looks okay.

h-assmann commented:
> $ umpf tag --flags upstreamstatus=insert

I'm not sure if we really want these flags as part of the umpf tag, since this effectively means that we will end up with two kinds of tags, one of which will not be (directly) usable with Yocto.

Unless we declare the usage of an upstreamstatus-flag as a general principle (in that case, the flag might become an opt-out parameter): Wouldn't it make more sense to insert the upstream-status when running "format-patch --bb"?

michaelolbrich commented:
The Upstream-Status is inserted during `umpf tag` so format-patch is too late. And doing this there is deliberate. I want it in the tag because otherwise the tag and the patches don't match.

And in practice, I've never seen a case where a tag is used for both Yocto and PTXdist so we just create patches Upstream-Status for Yocto and without for everything else.

Maybe umpf should refuse to `format-patch --bb` if the Upstream-Status is missing? @ejoerns , @OnkelUlla what do you think?

ejoerns commented:
> And in practice, I've never seen a case where a tag is used for both Yocto and PTXdist so we just create patches Upstream-Status for Yocto and without for everything else.

Agreed.

> Maybe umpf should refuse to `format-patch --bb` if the Upstream-Status is missing? @ejoerns , @OnkelUlla what do you think?

Using `--bb` seems to be more intuitive and well-known for umpf users than specifying the flag during tag creation already. Thus warning or erroring when having `--bb` but no upstream status could be a good way to enforce consistency here.

OnkelUlla commented:
We have to keep in mind that an upstream status is only required for patches that touch core-recipes. At the time being this is presumably the minority of patchseries as umpf tags are mostly created for kernel or barebox recipes in a customer's meta-layer. Hence it should not error out but at most warn about the possibly missing upstream status.

ejoerns commented:
A developer noted that they ran into issues with upstream status check (QA check `patch status`)  for custom layers. I have had another look now and indeed the check is enabled for oe-core only at the moment (now using the newly introduced layer overrides). Thus I have no idea how this would affect custom layers by default.

While it might still make sense to have this consistently in non-core layers (and the option to enable the QA check), I'd tend to agree that we should not enforce this at the moment.

michaelolbrich commented:
- I've added the Upstream-Status to the version patch
- Creating patches with `--bb` prints a message if no Upstream-Status was added, but will not fail
- Added a basic test.
